Improved Scrolling and Resizing for Tabs

just-the-benno opened the pull request:
I've worked on the scroll features of MudTab in conjunction with resizing. 

Currently, Mudtabs has a panel div item, which is as long as its children's sum. If it doesn't fit the screen, scrolling is necessary. Scrolling is done by setting translateX (or translateY) accordingly. 

The existing code has had some issues and errors where the state hasn't been updated accordingly. 

A major flaw where excessive javascript interop calls to get the size of elements. Besides, the inability to be updated when the tabs sizes (not the document) changes made some major changes necessary.

### New features
- [x] When a tab is selected, the viewport is centered around the item. Other items are displayed right and left if there are items and enough space 

### Fixes

- [x] The Disabled/Enabled states of the scrolling buttons are now updated accordingly 
- [x] Resizing hasn't taken into account that a single tab could have been resized
- [x] Resizing hasn't taken into account that only the parent container's size could change but not the document size
- [x] ActivePanelIndex hasn't been updated when a Tab has been removed 

### Improvements

- [x] Reduced calls to JS interop. Only once, while initializing, or as soon as a resize occurred, or if a tab is added
- [x] Unitests for scrolling, selecting items

### New Service: IResizeObserver

The ```IResizeListenerService``` is attached to the entire window/document's resize event, while the IResizeObserver is observing DOM elements. Hence, in the circumstances where the document size hasn't changed but the size of the DOM element, the IResizeObserver fires an event. Besides, the IResizeObserver has a cache for the ```bounding rectangle``` of each observed element. 

### Breaking changes

There are no breaking changes, except, changes in accessibility, But I assume (or hope) that these fields haven't been used before by anyone. 

Before
```C#
public ElementReference TabsContentSize
public List<MudTabPanel> Panels {get; set; } = new List<MudTabPanel>();
```

``` C#
private ElementReference _tabsContentSize
private List<MudTabPanel> _panels = new List<MudTabPanel>();
```

mikes-gh commented:
@just-the-benno this is great thank you.  We've been meaning to switch to resize observer for some time.  I'll try an do a full review in the next few days.

mikes-gh commented:
`JavascriptBasedResizeObserverFactory` can it be named `ResizeObserverFactory`
`JavascriptBasedResizeObserver` can it be named `ResizeObserver`

just-the-benno commented:
Hi @mikes-gh

It can be renamed. Should I do it? Or are you doing it during the review process?

mikes-gh commented:
I was seeking your opinion.  We don't like to be prescriptive but I feel in this case the naming could be simplified if you agree.

just-the-benno commented:
Agree 😸

mikes-gh commented:
I have done a scan review.
If you could commit the name change before I start review that would be best as the files will disappear on rename.

just-the-benno commented:
done

mikes-gh commented:
I ran your branch and all seems to functioning as expected great job.  Now it also works in the Iframe in tryMB so that proves ResizeObserver is doing its job.  However I am not familiar with the `IAsyncDispose` pattern and how it fits into the lifecycle of a Blazor component.  We can wait for @henon to have a look or you can educate me and I will continue the review.

A couple of things I'm not going to review line by line as its too messy.

1. You are using an old version of dev as your PR branch.  This means I am not sure exactly how your changes impact the current dev.  You also cant rebase on dev as you are already using it.  This will get messy as well if you want to track upstream/dev in your repo for future PR's rebase etc.  It's not easy to get out of now unless you are a git expert.  We can probably leave it and I can sort it out on the merge but it will make your life a lot easier if you always start new a new branch from an up to date `dev` for future PR's

2. It is uncommon to use c# types directly `Int32` `Double` `Boolean` etc.  In this repo we always use aliases https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/built-in-types.  Could you use aliases please as this is what the rest of the code base uses.

just-the-benno commented:
Thanks for the feedback and your time, too.

**Point 2:**
No problem, I can change it. I prefer, for instance, ```Int32``` instead of ``int``` because ```int``` is usually displayed with control structure syntax like a for-loop. Back in the days with C and even C++ ```int``` was actually very meaningful to the compile. It is not just a type, as it is in C#. So, to honor this progress, I deliberately (and proudly) choose Int32. 😃 However, your argument is valid and solid. 

**Point 1:**
Excellent point and definitely something I had to think about before starting working on it. I'll try my best to clean it up. 

The **IAsyncDispose** pattern is described [here](https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-disposeasync) and very similar to the synchronize dispose pattern. The method ```DisponseAsync``` is called by the RenderTree when an element is removed from the tree. The method is called before ```Dispose```.

The intention is to destroy the javascript ResizeObserver and free the occupied resources. This is a javascript interop call, and all javascript calls are asynchronous. The method ```InvokeVoidAsync``` is a wrapper around ```InvokeAsync<Object>``` and doesn't return any object. However, if the ```dispose``` method is finished before the javascript call has been executed, and ```TaskCanceledException ``` is thrown.  

With this approach and BlazorServer hosting model (not with Blazor WASM), there is a delay when navigated back from a page. However, it ensures that allocated resources (also like event handlers) are properly destroyed.

henon commented:
@mikes-gh my take is that IAsyncDisposable makes sense for the ResizeListener because it has to call JS in its dispose logic. For the TabPanel it does not really make sense and is only overcomplicated. I think it better be changed to normal IDispsoable logic.

just-the-benno commented:
@henon 

I can't entirely agree. That the method ```RemovePanel``` returning a task is important. RemovePanel leads to a JS call to unobserved the element. Hence, it is async.  Besides, it could lead to panel changes, so a call to ```StateHasChanged``` is needed.  This call should be wrapped inside ```InvokeAsync```, because a different thread could initiate the removal. Moreover, this call should be awaited to make sure that everything is executed in the expected sequence.

mikes-gh commented:
@henon The point about avoiding TaskCancelledExceptions and providing a deterministic tear down seems valid to me.  Having worked hard to avoid JS related exceptions due to object disposing before JS has completed I want to avoid introducing further exceptions.  Once you discard you lose control of the JS call.  In fact I think Steve Sanderson was promoting IAsyncDispose for this very situation on a aspnetcore issue I read. https://github.com/dotnet/aspnetcore/issues/13026#issuecomment-704859949

mikes-gh commented:
@just-the-benno Are you saying without awaiting the JSInterop in dispose there is a potential to not clean up unmanaged resources (memory leak)?

henon commented:
> @just-the-benno Are you saying without awaiting the JSInterop in dispose there is a potential to not clean up unmanaged resources (memory leak)?

I don't think so. But awaiting also doesn't hurt here except the slightly increased code complexity. 

Btw, what about the known problem that the element might be disposed before OnAfterRenderAsync is called? Might this problem hit us here?

just-the-benno commented:
@mikes-gh

In general, it is not a good practice to call any asynchronous functions inside a dispose method because you can never be sure that the method starts executing or finishes before the dispose is finished. Fire and Forget isn't (or shouldn't) be an option for dispose.

To answer your question, I start with the example of Tabs and then try to answer it in more general terms. 

Inside the javascript implementation of the MudResizeObserver, there is an associative array between an observer id (generated by the ResizeObserver) and the class handling to connect elements to the observer and disconnect from them. This class has a reference to the dotnet ResizeObserver. 

The javascript called inside the ```DisposeAsync``` method removes the binding in the associative array is deleted, and the javascript garbage collector can take care of the object. No reference back to the dotnet object. On the dotnet side, the DotNetReference is disposed of as well, so it could be garbage collected. 

So, in this case, without the call to js, we would have a memory leak inside javascript, and without disposing of the DotnetReference, we would have a memory leak inside .NET. In my experience, memory leaks inside Blazor are caused by not disposed DotnetReferences like inside the ```ResizeListenerService``` or not removed event listeners.

just-the-benno commented:
@henon  Do you have a test case for this scenario? The first connection to the observer happens in the ```OnAfterRenderAsync```. I guess that in the case an element is detached from a parent before this method is executed, the ```Context``` property of the ```ElementReference``` is (still) null. Only "valid" references with a context that is not null are observed, so there shouldn't be an issue. But, it is a guess. If this is a common thing, we should write a test case for it.

mikes-gh commented:
@henon

> Btw, what about the known problem that the element might be disposed before OnAfterRenderAsync is called? Might this problem hit us here?

We now do a null check on all elementReferences in the JS.

henon commented:
@just-the-benno 

ok, good, thanks for the thorough answers. @mikes-gh you may merge if you approve.

henon commented:
Oh, one more thing I want to discuss. The Breaking changes to invoke StateHasChanged asynchronously. I have researched StateHasChanged extensively and as I understand it, it sets a "dirty"-flag and gives the renderer a nudge. By awaiting StateHasChanged you gain absolutely nothing. To my knowledge the await will not wait for the render in this case.

So I don't think we should make these breaking changes.

mikes-gh commented:
@henon. I had a good look .  Which part of the API does this PR break please.  I couldn't find it.

just-the-benno commented:
@henon. I've seen a couple of examples where InvokeAsync has been awaited. It seems natural to me, especially regarding the name, but I've never intensively thought it through. Give me some days to research, playing around, and test it before accepting the PR. 

Another question is the ResizeObserver only "internal," or should it be open to the public. In the second case, a new page in the docs helps understand the usage, right?

henon commented:
Sure, no problem. I'll mark it as draft in the meantime. 

There is one case I know where it is crucial to call `InvokeAsync(StateHasChanged)`  instead of `StateHasChanged` and that is when you are not on the UI-thread. Another one might be when you don't want to immediately re-render. In all other cases `StateHasChanged` alone is actually better because otherwise it might cause a second render after the render which might already been scheduled due to parameter changes.

Of course it would be nice to document the ResizeObserver. For instance, we have ResizeListener documented here: https://mudblazor.com/components/hidden 
or it might fit in another section of this page https://mudblazor.com/features/breakpoints where I originally had the ResizeListener but it was moved to the hidden page. Or we make a new feature page dedicated to responsive UI, i.e. `features/responsive` and put ResizeListener and ResizeObserver there.

henon commented:
> @henon. I had a good look . Which part of the API does this PR break please. I couldn't find it.

One example is outlined in the description above by the author and it is hinted that there are more.

mikes-gh commented:
@henon @just-the-benno 
```
public async Task ActivatePanel(MudTabPanel panel){...}
```
Looks like this is the only breaking change? 

>the author and it is hinted that there are more.
I cant find them can you double check.

mikes-gh commented:
Can we forward a sync version of `ActivatePanel` to the async version using a discard and mark it obsolete?

just-the-benno commented:
@henon . Thank you so much for challenging my assumption, it has been a very interesting journey to find the answers and many surprises down the path.  I've made the appropriate changes. 

So there is no breaking change anymore, except from two changes of the accessibility of fields. I've updated the first post accordingly. 

@mikes-gh  I've rebased it to the current dev branch

@mikes-gh  I've changed the variables names from ```Int32``` to ```int``` and so on.

@henon  I like the idea of having a dedicated page as you mentioned like ```features/responsive```. I'm happy to provide the content regarding the ResizeObserver, but feeling not (yet) confident enough to write this page on my own.

To sum up, the pull request is ready for review, again :) 

And thank you so much for your honest input and questions.

mikes-gh commented:
@just-the-benno great news.  I'll have another look.  BTW you did a merge with dev which is not a rebase. You can see that from your commits including some that aren't yours. It doesn't matter as I will squash and merge this but just for your future info.

mikes-gh commented:
@just-the-benno in fact what I might do with your permission is rebase this on current dev because its showing 60 files to review?  That will mean you will have to reset your local branches though.  Is that OK?

just-the-benno commented:
@mikes-gh Agree. And I'm sorry for the mess. Next time, I start from the current dev branch

mikes-gh commented:
@just-the-benno Please review #1321 to see your changes are preserved and confirm.
I couldn't recover to your PR branch as it is dev and by default I only have push access to another users PR branches.

mikes-gh commented:
If you wish to do any significant further work on this I would

1.  Double check my PR has all your work
2.  Reset your dev `git reset --hard upstream/dev` 
3.  Create a new branch from your up to date `dev` and cherry pick the commit from my PR.

That will get you to a clean slate from which you can develop further if required.
